File: integration-tests/TestAPI-bindings/kotlin/src/test/kotlin/com/cricut/testapi/MethodTests.kt
package com.cricut.testapi

import kotlinx.coroutines.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import java.lang.Exception

@kotlinx.coroutines.ExperimentalCoroutinesApi
internal class MethodTests {
    @Test
    fun testStaticMethods() {
        assertEquals(123, Methods.staticGet)
        assertEquals(234, Methods.staticGetMethod())

        assertEquals(345, Methods.staticModifiable)
        Methods.staticModifiable = 987
        assertEquals(987, Methods.staticModifiable)

        assertEquals(567, Methods.staticStored)
        Methods.staticStored = 765
        assertEquals(765, Methods.staticStored)
    }

    @Test
    fun testInstanceMethods() {
        val instance = Methods.create()

        assertEquals(1234, instance.instanceGet)
        assertEquals(2345, instance.instanceGetMethod())

        assertEquals(3456, instance.instanceModifiable)
        instance.instanceModifiable = 9876
        assertEquals(9876, instance.instanceModifiable)

        assertEquals(5678, instance.instanceStored)
        instance.instanceStored = 7654
        assertEquals(7654, instance.instanceStored)
    }

    @Nested
    inner class AsyncInstanceTests {
        @Test
        fun testAsyncInstanceMethods() {
            runBlocking {
                val value: Long = Methods.create().async42()
                assertEquals(42, value)
            }
        }

        @Test
        fun testAsyncYieldingFunctionCall() {
            runBlocking {
                Methods.create().asyncYield()
            }
        }

        @Test
        fun testAsyncSleepFunctionCal() {
            runBlocking {
                Methods.create().asyncSleep()
            }
        }

        @Test
        fun testAsyncVoidFunctionCall() {
            runBlocking {
                Methods.create().asyncVoid()
            }
        }

        @Test
        fun testAsyncCallbackFunctionCall0() {
            runBlocking {
                var threw = false
                var value: Long = 42
                var ran = false
                try {
                    value = Methods.create().asyncCallbackFunc0 {
                        ran = true
                        42
                    }
                } catch (e: Exception) {
                    threw = true
                }
                assertEquals(false, threw)
                assertEquals(42, value)
                assertEquals(true, ran)
            }
        }

        @Test
        fun testAsyncCallbackFunctionCallThrow() {
            assertThrows(Error::class.java) {
                runBlocking {
                    Methods.create().asyncCallbackFunc0 { throw java.lang.Error() }
                }
            }
        }

        @Test
        fun testAsyncDoubleFunctionCall() {
            runBlocking {
                val value = Methods.create().asyncDouble(1.0)
                assertEquals(2.0, value)
            }
        }

        @Test
        fun testAsyncMultipleArgsFunctionCall() {
            runBlocking {
                assertEquals(3, Methods.create().asyncMultipleArgs(1) { 2 })
            }
        }

        @Test
        fun testAsyncThrowingFunctionCall() {
            assertThrows(Error::class.java) {
                runBlocking {
                    Methods.create().asyncThrowing()
                }
            }
        }
    }

    @Nested
    inner class AsyncStaticTests {
        @Test
        fun testAsyncInstanceMethods() {
            runBlocking {
                val value: Long = Methods.staticAsync42()
                assertEquals(42, value)
            }
        }

        @Test
        fun testAsyncYieldingFunctionCall() {
            runBlocking {
                Methods.staticAsyncYield()
            }
        }

        @Test
        fun testAsyncSleepFunctionCal() {
            runBlocking {
                Methods.staticAsyncSleep()
            }
        }

        @Test
        fun testAsyncVoidFunctionCall() {
            runBlocking {
                Methods.staticAsyncVoid()
            }
        }

        @Test
        fun testAsyncCallbackFunctionCall0() {
            runBlocking {
                var threw = false
                var value: Long = 42
                var ran = false
                try {
                    value = Methods.staticAsyncCallbackFunc0 {
                        ran = true
                        42
                    }
                } catch (e: Exception) {
                    threw = true
                }
                assertEquals(false, threw)
                assertEquals(42, value)
                assertEquals(true, ran)
            }
        }

        @Test
        fun testAsyncCallbackFunctionCallThrow() {
            assertThrows(Error::class.java) {
                runBlocking {
                    Methods.staticAsyncCallbackFunc0 { throw java.lang.Error() }
                }
            }
        }

        @Test
        fun testAsyncDoubleFunctionCall() {
            runBlocking {
                val value = Methods.staticAsyncDouble(1.0)
                assertEquals(2.0, value)
            }
        }

        @Test
        fun testAsyncMultipleArgsFunctionCall() {
            runBlocking {
                assertEquals(3, Methods.staticAsyncMultipleArgs(1) { 2 })
            }
        }

        @Test
        fun testAsyncThrowingFunctionCall() {
            assertThrows(Error::class.java) {
                runBlocking {
                    Methods.staticAsyncThrowing()
                }
            }
        }
    }
}

File: integration-tests/TestAPI-bindings/c-sharp/Cricut.TestAPI.Tests/MethodTests.cs
using System;
using System.Threading.Tasks;
using FluentAssertions;
using Xunit;

namespace Cricut.TestAPI.Tests {
    public class MethodTests {
        [Fact]
        void testStaticMethods() {
            Assert.Equal(123, Methods.GetStaticGet());
            Assert.Equal(234, Methods.GetStaticGetMethod());

            Assert.Equal(345, Methods.GetStaticModifiable());
            Methods.SetStaticModifiable(987);
            Assert.Equal(987, Methods.GetStaticModifiable());

            Assert.Equal(567, Methods.StaticStored);
            Methods.StaticStored = 765;
            Assert.Equal(765, Methods.StaticStored);
        }

        [Fact]
        void testInstanceMethods() {
            var instance = Methods.Create();

            Assert.Equal(1234, instance.GetInstanceGet());
            Assert.Equal(2345, instance.GetInstanceGetMethod());

            Assert.Equal(3456, instance.GetInstanceModifiable());
            instance.SetInstanceModifiable(9876);
            Assert.Equal(9876, instance.GetInstanceModifiable());

            Assert.Equal(5678, instance.InstanceStored);
            instance.InstanceStored = 7654;
            Assert.Equal(7654, instance.InstanceStored);
        }

        public class AsyncInstanceTests {
            [Fact]
            async Task TestAsyncFunctionCall() {
                var value = await Methods.Create().Async42();
                Assert.Equal(42, value);
            }

            [Fact]
            async Task TestAsyncYieldingFunctionCall() {
                await Methods.Create().AsyncYield();
            }

            [Fact]
            async Task TestAsyncSleepFunctionCall() {
                await Methods.Create().AsyncSleep();
            }

            [Fact]
            async Task TestAsyncVoidFunctionCall() {
                await Methods.Create().AsyncVoid();
            }

            [Fact]
            async Task TestAsyncCallbackFunctionCall0() {
                var threw = false;
                nint value = 42;
                var ran = false;
                try {
                    value = await Methods.Create().AsyncCallbackFunc0(() => {
                        ran = true;
                        return Task.FromResult<nint>(42);
                    });
                } catch {
                    threw = true;
                }
                Assert.False(threw);
                Assert.Equal(42, value);
                Assert.True(ran);
            }

            [Fact]
            async Task TestAsyncCallbackFunctionCallThrow() {
                var theException = new Exception("asyncErr");
                var exception = await Assert.ThrowsAsync<Exception>(() =>
                    Methods.Create().AsyncCallbackFunc0(() => throw theException)
                );
                Assert.Equal(theException, exception);
            }

            [Fact]
            async Task TestAsyncDoubleFunctionCall() {
                var value = await Methods.Create().AsyncDouble(1.0);
                Assert.Equal(2.0, value);
            }

            [Fact]
            async Task TestAsyncMultipleArgsFunctionCall() {
                Assert.Equal(3, await Methods.Create().AsyncMultipleArgs(1, () => Task.FromResult<nint>(2)));
            }

            [Fact]
            async Task TestAsyncThrowingFunctionCall() {
                var exception = await Assert.ThrowsAsync<Exception>(async () => await Methods.Create().AsyncThrowing());
                Assert.Equal("TheMethodError()", exception.Message);
            }
        }
    }

    public class AsyncStaticTests {
        [Fact]
        async Task TestAsyncFunctionCall() {
            var value = await Methods.StaticAsync42();
            Assert.Equal(42, value);
        }

        [Fact]
        async Task TestAsyncYieldingFunctionCall() {
            await Methods.StaticAsyncYield();
        }

        [Fact]
        async Task TestAsyncSleepFunctionCall() {
            await Methods.StaticAsyncSleep();
        }

        [Fact]
        async Task TestAsyncVoidFunctionCall() {
            await Methods.StaticAsyncVoid();
        }

        [Fact]
        async Task TestAsyncCallbackFunctionCall0() {
            var threw = false;
            nint value = 42;
            var ran = false;
            try {
                value = await Methods.StaticAsyncCallbackFunc0(() => {
                    ran = true;
                    return Task.FromResult<nint>(42);
                });
            } catch {
                threw = true;
            }
            Assert.False(threw);
            Assert.Equal(42, value);
            Assert.True(ran);
        }

        [Fact]
        async Task TestAsyncCallbackFunctionCallThrow() {
            var theException = new Exception("asyncErr");
            var exception = await Assert.ThrowsAsync<Exception>(() =>
                Methods.StaticAsyncCallbackFunc0(() => throw theException)
            );
            Assert.Equal(theException, exception);
        }

        [Fact]
        async Task TestAsyncDoubleFunctionCall() {
            var value = await Methods.StaticAsyncDouble(1.0);
            Assert.Equal(2.0, value);
        }

        [Fact]
        async Task TestAsyncMultipleArgsFunctionCall() {
            Assert.Equal(3, await Methods.StaticAsyncMultipleArgs(1, () => Task.FromResult<nint>(2)));
        }

        [Fact]
        async Task TestAsyncThrowingFunctionCall() {
            var exception = await Assert.ThrowsAsync<Exception>(async () => await Methods.StaticAsyncThrowing());
            Assert.Equal("TheMethodError()", exception.Message);
        }
    }
}
